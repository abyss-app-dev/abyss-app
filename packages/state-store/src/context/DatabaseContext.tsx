import type { SQliteClient } from '@abyss/records';
import { createContext, type ReactNode, useContext } from 'react';

interface DatabaseContextType {
    database: SQliteClient | null;
}

const DatabaseContext = createContext<DatabaseContextType>({ database: null });

interface DatabaseProviderProps {
    database: SQliteClient;
    children: ReactNode;
}

export function DatabaseProvider({ database, children }: DatabaseProviderProps) {
    return <DatabaseContext.Provider value={{ database }}>{children}</DatabaseContext.Provider>;
}

export function useDatabase(): SQliteClient {
    const context = useContext(DatabaseContext);
    if (!context.database) {
        throw new Error('useDatabase must be used within a DatabaseProvider');
    }
    return context.database;
}
