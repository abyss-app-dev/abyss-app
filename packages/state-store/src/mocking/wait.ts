export function wait(ms = 50) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
