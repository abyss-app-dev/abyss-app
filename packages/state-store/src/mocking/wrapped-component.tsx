import type { StateRecorder } from './state-recorder';

export function TestRecorder<T>({ functionToTest, recorder }: { functionToTest: () => T; recorder: StateRecorder }) {
    recorder.record(functionToTest());
    return <>content</>;
}
