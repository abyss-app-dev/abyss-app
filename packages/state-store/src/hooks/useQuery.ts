import type { SQliteClient } from '@abyss/records';
import { useCallback, useState } from 'react';
import { useDatabase } from '../context/DatabaseContext';
import { useOnce } from './useOnce';

export type DbQuery<T> = (database: SQliteClient) => Promise<T>;
export type Dependencies = unknown[];

export function useQuery<T>(callback: DbQuery<T>, dependencies: Dependencies = []) {
    const database = useDatabase();
    const [data, setData] = useState<T | null>(null);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState<Error | null>(null);

    const fetchData = useCallback(async () => {
        try {
            setLoading(true);
            const result = await callback(database);
            setData(result);
        } catch (error) {
            setError(error as Error);
            setData(null);
        } finally {
            setLoading(false);
        }
    }, [callback, database, ...dependencies]);

    // Initial query called 1x
    useOnce(fetchData, [callback, database, ...dependencies]);

    return { data, loading, error, refetch: fetchData, setData };
}
