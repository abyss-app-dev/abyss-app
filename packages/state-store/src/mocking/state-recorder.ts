import { render } from '@testing-library/react';
import type { ReactNode } from 'react';
import { expect } from 'vitest';
import { wait } from './wait';

export class StateRecorder {
    private states: unknown[] = [];

    record(state: unknown): void {
        this.states.push(state);
    }

    async delayedResolve(): Promise<unknown[]> {
        await wait();
        return this.states;
    }

    getStates(): unknown[] {
        return this.states;
    }

    getLatestState(): unknown {
        return this.states[this.states.length - 1];
    }

    clear(): void {
        this.states.length = 0;
    }

    async renderToSnapshot(component: ReactNode): Promise<void> {
        render(component);
        expect(await this.delayedResolve()).toMatchSnapshot();
    }
}
