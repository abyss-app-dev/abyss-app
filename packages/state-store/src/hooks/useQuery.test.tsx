import { buildTestDB, type SQliteClient, SqliteTable } from '@abyss/records';
import { useEffect, useState } from 'react';
import { beforeEach, describe, test } from 'vitest';
import { DatabaseProvider } from '../context/DatabaseContext';
import { StateRecorder } from '../mocking/state-recorder';
import { TestRecorder } from '../mocking/wrapped-component';
import { useQuery } from './useQuery';

describe('useQuery hook', () => {
    let db: SQliteClient;
    let recorder: StateRecorder;

    beforeEach(async () => {
        db = await buildTestDB();
        recorder = new StateRecorder();
    });

    test('load data from database using a query', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const queryResults = useQuery(db => db.tables[SqliteTable.settings].default());
                        return queryResults;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('query results in error', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const queryResults = useQuery(() => {
                            throw new Error('Query failed');
                        });
                        return queryResults;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('query has dependencies that change', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [dependency, setDependency] = useState(1);
                        const queryResults = useQuery(() => Promise.resolve(dependency), [dependency]);
                        useEffect(() => {
                            setTimeout(() => setDependency(2));
                        }, [dependency]);
                        return queryResults;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('query has handler which changes', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [dependency, setDependency] = useState(1);
                        const queryResults = useQuery(() => Promise.resolve(dependency), [dependency]);
                        useEffect(() => {
                            setTimeout(() => setDependency(2));
                        }, [dependency]);
                        return queryResults;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('call refetch to reload data', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const queryResults = useQuery(() => Promise.resolve('initial data'));
                        useEffect(() => {
                            if (queryResults.data === 'initial data') {
                                setTimeout(() => queryResults.refetch());
                            }
                        }, [queryResults.data, queryResults.refetch]);
                        return queryResults;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('call setData to manually update state', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const queryResults = useQuery(() => Promise.resolve('initial data'));
                        useEffect(() => {
                            if (queryResults.data === 'initial data') {
                                setTimeout(() => queryResults.setData('manually set data'));
                            }
                        }, [queryResults.data, queryResults.setData]);
                        return queryResults;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('transition from data to failure', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [shouldFail, setShouldFail] = useState(false);
                        const queryResults = useQuery(() => {
                            if (shouldFail) {
                                throw new Error('Query now fails');
                            }
                            return Promise.resolve('successful data');
                        }, [shouldFail]);

                        useEffect(() => {
                            if (queryResults.data === 'successful data') {
                                setTimeout(() => setShouldFail(true));
                            }
                        }, [queryResults.data]);

                        return queryResults;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('transition from failure to data', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [shouldFail, setShouldFail] = useState(true);
                        const queryResults = useQuery(() => {
                            if (shouldFail) {
                                throw new Error('Initial failure');
                            }
                            return Promise.resolve('recovered data');
                        }, [shouldFail]);

                        useEffect(() => {
                            if (queryResults.error) {
                                setTimeout(() => setShouldFail(false));
                            }
                        }, [queryResults.error]);

                        return queryResults;
                    }}
                />
            </DatabaseProvider>
        );
    });
});
