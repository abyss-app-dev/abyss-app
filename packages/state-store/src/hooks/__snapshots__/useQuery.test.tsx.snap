// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useQuery hook > call refetch to reload data 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "initial data",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "initial data",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useQuery hook > call setData to manually update state 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "initial data",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "manually set data",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useQuery hook > load data from database using a query 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748749550396,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749550396,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useQuery hook > query has dependencies that change 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": 1,
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": 1,
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": 1,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": 2,
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useQuery hook > query has handler which changes 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": 1,
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": 1,
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": 1,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": 2,
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": 2,
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useQuery hook > query results in error 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": [Error: Query failed],
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useQuery hook > transition from data to failure 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "successful data",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "successful data",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": [Error: Query now fails],
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useQuery hook > transition from failure to data 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": [Error: Initial failure],
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": [Error: Initial failure],
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": [Error: Initial failure],
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "recovered data",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;
